sdks/python/apache_beam/runners/direct/transform_evaluator.py

```diff
@@ -524,15 +529,19 @@ def process_element(self, element):
   def finish_bundle(self):
     # TODO(altay): Do not wait until the last bundle to write in a single shard.
     if self._is_final_bundle:
-      if isinstance(self._sink, io.fileio.NativeTextFileSink):
-        assert self._sink.num_shards in (0, 1)
-        if self._sink.shard_name_template:
-          self._sink.file_path += '-00000-of-00001'
-          self._sink.file_path += self._sink.file_name_suffix
-      self._sink.pipeline_options = self._evaluation_context.pipeline_options
-      with self._sink.writer() as writer:
-        for v in self.state:
-          writer.Write(v.value)
+      if not self._has_already_produced_output:
```

Comment:
I understand this is a surgical fix, but it is worth a comment that three nested conditionals is getting a bit hard to read, especially with the distance between them. I'd at least drop the `not` and put the short branch first so the `if` and `else` are more adjacent.